### ISSUE TITLE
Ability to switch graphics depending of the power source 

### ISSUE THREAD

**Issue by heaven** [rugarciap/Turbo-Boost-Switcher#42]
Apple allows to adjust some settings depending on the power source, but not to switch graphics. I would like to use discrete graphic working from the power adapter.
<img width="780" alt="screen shot 2017-08-29 at 13 46 37" src="https://user-images.githubusercontent.com/126732/29817398-85a794d0-8cc0-11e7-9b72-ae37f44bb8b6.png">

**Comment by rugarciap** [comment]
Hi!

I know this is old, but could you please elaborate a little more? Thanks!

**Comment by heaven** [comment]
Hi, sure. The problem is I don't have a setting to prefer discrete GPU when on power and integrated when working from the battery. On the screenshot, you can see the graphics switching option was moved outside the battery/power adapter tabs.

**Comment by heaven** [comment]
Not related to turbo boost, but I was thinking this could be a great feature that is missing in mac os.

**Comment by ideologysec** [comment]
Seems also like, just as you can enable/disable TurboBoost on battery to save power, @heaven is asking to be able to automatically use the dedicated GPU when connected to power.

So, on battery, switches automatically/prefers integrated GPU. when connected to power, the dedicated GPU is used.

**Comment by heaven** [comment]
Yeah, something like that. Mac OS doesn't have the ability to set the GPU preference. You can see on the screenshot above the graphics switching setting is out of the scope and there's no way to set up the GPU preference based on the modes.

Often, when working from the battery, during long trips, waiting in airports or working from a plane I need the maximum battery life. And something like Google Chrome forces discrete GPU and dries the battery in a few hours.

I reported this issue to Apple but they don't care. So I was thinking it would be an awesome feature if there was a checkbox that would disable discrete GPU when working from the battery:
<img width="380" alt="Screen shot 2019-11-08 at 09 49 06" src="https://user-images.githubusercontent.com/126732/68459086-0df1c400-020d-11ea-81a7-07f50a03a719.png">

I am not sure if this is possible to implement, this is just a setting that I really miss in Mac OS.

**Comment by jb510** [comment]
For years I used https://github.com/codykrieger/gfxCardStatus for forcing the discrete GPU off to extend battery life and reduce heat generation on my MBPs. It was awesome, but @codykrieger went to work for Apple and stopped supporting it since it used private apis. I’m fairly certain it stopped working all together in newer MBPs or with 10.14. 

Above is a bit confusing.  There is no real performance benefit to forcing the dGPU on.  Auto switching works great. There is a huge power and heat savings to forcing it off however, with minimal impact to most non-intensive workloads (ie, browsing, coding, typing, not rendering).

It would be a welcome related feature to disable dGPU switching along with disabling CPU turbo If that even still possible.

**Comment by ideologysec** [comment]
It's no longer possible to force integrated-only GPU; that changed back in High Sierra I believe.

**Comment by jb510** [comment]
> It's no longer possible to force integrated-only GPU; that changed back in High Sierra I believe.

I stand corrected, 10.13 not 14

**Comment by codykrieger** [comment]
@jb510 Well, that's one version of events, I suppose. 😛

In actuality, when I was at Apple, I had special dispensation to continue supporting gfxCardStatus to a degree.

I'm pretty sure the root cause of the breakage comes down to changes in AppleGraphicsControl.kext, which is how gfxCardStatus interfaces with the graphics muxing system. The way gfxCardStatus speaks to AppleGraphicsControl is based on some reverse-engineering originally done by a semi-anonymous poster on the MacRumors forums back in 2010. With some effort, it may be possible to fix the brokenness introduced in the 10.13-ish timeframe.

While I now have the know-how to do this kind of reverse-engineering myself, ultimately, I don't think it's worthwhile. The vast majority of remaining gfxCardStatus users are folks trying to prolong the life of older machines with defective hardware—which is all well and good—but the automatic graphics switching in modern macOS is pretty damn good, and the hardware seems to be more reliable (y'know, aside from the keyboards).

That being the case, gfxCardStatus seems a lot less necessary now than it was in 2010. 🤷‍♀

**Comment by jb510** [comment]
Cody thank you for that!   

I agree gfx switching is way better these days. Interesting people using it to get around broken hardware, I’d never thought of that. 

 I’m an edge case in that I often use an external usb monitor while still on battery. That would force the dGPU on and eat more power than really need.  But that was years ago and I have no idea yet if my 3 day old 16” will behave that way.  

Even then an external usb-pd battery has solved 99% of my unplugged all day power needs.  

Thanks again for the awesomeness that was/is Gfxswitcher.

**Comment by jb510** [comment]
Oh! And sorry about the confusion.  I clearly misremembered whatever it was I read when you announced going to work for Apple.  Maybe it was just a similar story someone else shared.

**Comment by codykrieger** [comment]
@jb510 Heh, no worries.

**Comment by zllovesuki** [comment]
On my 2019 16-inch MacBook Pro running macOS 10.15.3 Catalina, this still works to switch your graphics automatically based on your power source:

Switch to discrete graphics when on AC Power:
`sudo pmset -c gpuswitch 1`

Switch to integrated graphics when on Charger:
`sudo pmset -cbgpuswitch 0`

Or an one-liner:
`sudo pmset -c gpuswitch 1 && sudo pmset -b gpuswitch 0`

Explanation:
`pmset` changes your power profile depending on power source (`-c` for charging, `-b` for battery), `1` for discrete graphics, and `0` for integrated graphics.

---

To reset back to default behavior:
`sudo pmset -a gpuswitch 2`

Explanation:
`-a` for both AC and Battery.

---

Using this `pmset` in combination with Turbo Boost Switcher, running on battery is extremely cool and it saves on battery greatly.

Caveat:
If you are running something that specifically uses the discrete graphics, the application/script will still uses the discrete without the OS switching to the discrete graphics (as evident by running Keras while the system reports it is using integrated graphics, yet the discrete GPU is using 50W of power reported by iStat Menus).

**Comment by mjturner** [comment]
Bit belated, but I've found that [gSwitch](https://github.com/CodySchrank/gSwitch) works well as a replacement for gfxCardStatus. It possibly doesn't fit others' use cases, but I use it to force discrete graphics to avoid the annoying integrated/discrete switch when using certain applications (eg, when browsing an image-heavy site in Firefox).

**Comment by codykrieger** [comment]
@mjturner If that's all you want to do, you don't need anything like gfxCardStatus or gSwitch. Just disable _Automatic graphics switching_ in System Preferences > Energy Saver. That'll keep the discrete GPU on all the time.

**Comment by poige** [comment]
> On my 2019 16-inch MacBook Pro running macOS 10.15.3 Catalina, this still works to switch your graphics automatically based on your power source:
> 
> Switch to discrete graphics when on AC Power: `sudo pmset -c gpuswitch 1`
> 
> Switch to integrated graphics when on battery: `sudo pmset -b gpuswitch 0`
> 
> Or an one-liner: `sudo pmset -c gpuswitch 1 && sudo pmset -b gpuswitch 0`
> 
> Explanation: `pmset` changes your power profile depending on power source (`-c` for charging, `-b` for battery), `1` for discrete graphics, and `0` for integrated graphics.

There's another explanation @zllovesuki : _If it doesn't fail when you run it, it doesn't mean it worked as you expected_ (and persuaded others). It's simply just another way to the same knobs as with GUI and it won't make macOS handle it any smarter then its GUI.

**Comment by rugarciap** [comment]
Hi, 

Now that I've released another app called Power Mode Switcher ([www.pmswitcher.com](www.pmswitcher.com)), this could fit in that app better than inside Turbo Boost Switcher. Would that make sense to you? :)

**Comment by poige** [comment]
As to me TBSw is still appropriate app to add the feature — switching turbo boost on/off for a CPU is very close to changing between int/ext GPUs, kinda switching graphics boost on/off. And of course it should be just a different setting, not the all-in big on/off knob.
